### ISSUE TITLE
Very poor results with feature extraction & comparison

### ISSUE THREAD

**Issue by brr53** [ultralytics/ultralytics#5888]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I'm trying to do feature extraction with yolov8 and use cosine similarity to find similar images (classification)

However, I'm getting really really terrible results (accuracy). The similarities seem almost arbitrary. Similar images get lower similarity scores than very different images. Even much smaller Resnet18 model has significantly better results. What am I doing wrong? Any advice?

```import os
import torch
from torchvision import transforms
from PIL import Image
from ultralytics import YOLO
from torch.nn import Sequential, functional as F

# Load YOLO model and remove the last two layers
yolo = YOLO("yolov8l-cls.pt")
model = Sequential(*list(yolo.model.model.children())[:-2])
model.eval()

# Define preprocessing strategy (RGB)
preprocess = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

def get_features(img_path):
    img = Image.open(img_path).convert('RGB')
    x = preprocess(img).unsqueeze(0)
    with torch.no_grad():
        features = model(x)
    features = features.view(-1) # convert to 1d
    return features

def get_all_images_from_folder(folder_path):
    image_files = [f for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))]
    return [os.path.join(folder_path, f) for f in image_files if f.lower().endswith(('.png', '.jpg', '.jpeg'))]

folder_path = r'C:\Admin\animals'
image_paths = get_all_images_from_folder(folder_path)
image_features = {img_path: get_features(img_path) for img_path in image_paths}

# Create a dictionary to hold similarities for each image
similarity_dict = {os.path.basename(img_path): [] for img_path in image_paths}

# Calculate similarities and store
for i, img_path1 in enumerate(image_paths):
    for j, img_path2 in enumerate(image_paths):
        if i < j:
            similarity = F.cosine_similarity(image_features[img_path1].unsqueeze(0), image_features[img_path2].unsqueeze(0))
            similarity_value = similarity.item()
            similarity_dict[os.path.basename(img_path1)].append((os.path.basename(img_path2), similarity_value))
            similarity_dict[os.path.basename(img_path2)].append((os.path.basename(img_path1), similarity_value))

# Print results
for target_image, similarities in similarity_dict.items():
    sorted_similarities = sorted(similarities, key=lambda x: x[1], reverse=True)
    print(f"\nSimilarities for {target_image}:")
    for img, sim in sorted_similarities:
        print(f"  Similarity between {target_image} and {img}: {sim}")
```

**Comment by captainfffsama** [comment]
maybe use a L2 normalization  after get feature can get a better result

**Comment by glenn-jocher** [comment]
@brr53 hi there, and thanks for using our YOLOv8 model for your tasks! You mentioned you're having difficulties with feature extraction and comparison, and I'd like to suggest a few potential strategies to improve your results.

1. **Layer selection**: It's important to ensure the layer you're extracting features from is suitable for your specific task. Higher layers in the network usually contain more abstract features that may not be as useful for image similarity tasks as lower or mid-level features. You might consider experimenting with which layer's outputs you're using for the feature extraction.

2. **Normalization**: As you've mentioned using cosine similarity for measuring image similarity, applying L2 normalization to your extracted features can be beneficial. L2 normalization scales the features so they have a consistent scale, which can make the cosine similarity more meaningful. 

3. **Data variability**: If your dataset contains a wide variety of images, the model might struggle to match similarity correctly. In this case, you might want to consider training the model with a custom dataset more reflective of the kinds of images you're working with.

I hope these suggestions are helpful for you. Good luck with your project, and please don't hesitate to reach out if you have more questions.

**Comment by brr53** [comment]
Thank you for your suggestions @glenn-jocher .

1. I've tried removing the 1st, 2nd, 3rd, layer. They did not offer any accurate results.
2. I've added L2 normalization to the features by adjusting the code to the following:
    ```def get_features(img_path):
    img = Image.open(img_path)
    x = preprocess(img).unsqueeze(0)
    with torch.no_grad():
        features = model(x)
    features = features.view(-1) # convert to 1d
    features = F.normalize(features, dim=0)
    return features```
    
    It did not improve the accuracy of the similarities.
3. I'm using the same images used in training of the model.

What should I do? I cannot get accurate cosine similarity with Yolov8 Classification. Would love to use Yolov8.
